library/core/src/ptr/const_ptr.rs

```diff
@@ -1347,16 +1366,26 @@ impl<T: ?Sized> *const T {
     /// # Panics
     ///
     /// The function panics if `align` is not a power-of-two (this includes 0).
+    // #[cfg(not(bootstrap))] -- Calling this function in a const context from the bootstrap
+    // compiler will always return false.
     #[must_use]
     #[inline]
     #[unstable(feature = "pointer_is_aligned", issue = "96284")]
-    pub fn is_aligned_to(self, align: usize) -> bool {
-        if !align.is_power_of_two() {
-            panic!("is_aligned_to: align is not a power-of-two");
+    #[rustc_const_unstable(feature = "const_pointer_is_aligned", issue = "none")]
+    pub const fn is_aligned_to(self, align: usize) -> bool {
+        assert!(align.is_power_of_two(), "is_aligned_to: align is not a power-of-two");
+
+        #[inline]
+        fn runtime(ptr: *const u8, align: usize) -> bool {
+            ptr.addr() & (align - 1) == 0
+        }
+
+        const fn comptime(ptr: *const u8, align: usize) -> bool {
+            ptr.align_offset(align) == 0
         }
 
-        // Cast is needed for `T: !Sized`
-        self.cast::<u8>().addr() & align - 1 == 0
+        // SAFETY: `ptr.align_offset(align)` returns 0 if and only if the pointer is already aligned.
+        unsafe { intrinsics::const_eval_select((self.cast::<u8>(), align), comptime, runtime) }
```

Comment:
can we always invoke `ptr.align_offset(align) == 0` even at runtime? Or is that a performance concern?

Reply:
There is a small performance penalty: [Goldbolt link](https://rust.godbolt.org/z/Y4aExn36x)

Reply:
Ah, please leave a comment to that regard

Reply:
I would not call the performance penalty small. `*const u8` is easy mode here, you can see the code size explode if you make the pointee type `u16`, and for `u32` and larger `align_offset` is not even inlined.

Reply:
For the purpose of `is_aligned` we can just cast the pointer to `*const u8`, because we only care if the offset is zero or not zero. (And we do this cast already anyway to deal with fat pointers.)

Reply:
It seems inlined to me in https://rust.godbolt.org/z/b7n4MPGdf... But the difference is quite staggering:

```nasm
example::is_aligned_to_old_unchecked:
        dec     rsi
        test    rsi, rdi
        sete    al
        ret

example::is_aligned_to_new_unchecked:
        lea     r8, [rsi - 1]
        test    sil, 3
        je      .LBB1_1
        bsf     rax, rsi
        cmp     rax, 2
        mov     ecx, 2
        cmovb   rcx, rax
        mov     edx, -1
        shl     edx, cl
        not     edx
        mov     rax, -1
        test    edx, edi
        je      .LBB1_4
.LBB1_10:
        test    rax, rax
        sete    al
        ret
.LBB1_1:
        mov     rax, -1
        test    dil, 3
        jne     .LBB1_10
        add     r8, rdi
        neg     rsi
        and     rsi, r8
        sub     rsi, rdi
        shr     rsi, 2
        mov     rax, rsi
        test    rax, rax
        sete    al
        ret
.LBB1_4:
        shr     rsi, cl
        mov     r10d, r8d
        and     r10d, 4
        shr     r10, cl
        and     rdi, r8
        shr     rdi, cl
        lea     r8, [rsi - 1]
        mov     r9, rsi
        sub     r9, rdi
        mov     rax, r10
        shr     rax
        lea     rdi, [rip + .L__unnamed_1]
        movzx   edi, byte ptr [rax + rdi]
        cmp     rsi, 17
        jae     .LBB1_6
        mov     rax, rdi
        jmp     .LBB1_9
.LBB1_6:
        mov     rcx, r10
        imul    rcx, rdi
        mov     eax, 2
        sub     rax, rcx
        imul    rax, rdi
        cmp     rsi, 257
        jb      .LBB1_9
        mov     edi, 256
.LBB1_8:
        imul    rdi, rdi
        mov     rcx, rax
        imul    rcx, r10
        mov     edx, 2
        sub     rdx, rcx
        imul    rax, rdx
        cmp     rdi, rsi
        jb      .LBB1_8
.LBB1_9:
        and     rax, r8
        imul    rax, r9
        and     rax, r8
        test    rax, rax
        sete    al
        ret

.L__unnamed_1:
        .ascii  "\001\013\r\007\t\003\005\017"
```

Reply:
Note that align_offset is also large enough to never get inlined (even for u8) on `-Copt-level=s` (and probably `z` too). And we definitely don't want to `#[inline(always)]` it due to how much code it can generate in some cases.

Reply:
I found out that `align_offset == 0` does get optimized to the old `is_aligned_to` impl with `opt-level=1/2/3/s/z` if you cast the pointer to `*const ()` and `#[inline]` the `align_offset` method on pointers (not the big freestanding function): https://rust.godbolt.org/z/Kd98b9jvM

But the "optimized for size" code is still larger than the optimized for speed one, because it keeps the dead assembly for `align_offset` around.